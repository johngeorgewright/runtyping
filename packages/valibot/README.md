# @runtyping/valibot

Generate valibot from static types & JSON schema.

## Instructions

### Install

```
npm install valibot
npm install -D @runtyping/valibot
```

### Use from the command line

1. Create a file, in the root of your project, called "runtyping.yml".

   ```yaml
   # runtyping.yml

   targetFile: src/other-runtypes.ts # The file to create
   runtypeFormat: {type}Rt # Optional: use a custom name format for the created runtype
   typeFormat: {type}Type  # Optional: use a custom name format for the created type
   sourceTypes:
     exportStaticType: true # Optional: export static types as well (true by default)
     file: src/types.ts # The file where your type lives
     type: Foo # The type you want to convert to a runtype
   ```

   You can also specify a list of target files, if you want to create more than one:

   ```yaml
   # runtyping.yml

   - targetFile: src/other-runtypes.ts
     sourceTypes:
       file: src/types.ts
       type: Foo

   - targetFile: src/runtypes.ts
     sourceTypes:
       # Source types can also be a list
       - file: src/types.ts
         type: Foo

       - file: json/my-json-schema.json # You can even use JSON schema files!!
         type: [ExampleType, AnotherExampleType] # You may use an array of types
   ```

1. Then run: `npx runtyping`

### Use from a script

Basic example:

```ts
import { Generator } from '@runtyping/valibot'

const generator = new Generator({
  targetFile: 'src/runtypes.ts',
  // optional: runtypeFormat / typeFormat (see above)
})

generator
  .generate([
    { file: 'src/types.ts', type: 'Foo' },
    { file: 'json/my-json-schema.json', type: 'ExampleType' },
  ])
  .then((file) => file.save())
```

#### Passing a custom tsconfig file

```ts
import { Generator } from '@runtyping/valibot'
const generator = new Generator({
  targetFile: 'src/runtypes.ts',
  tsConfigFile: '/path/to/tsconfig.json',
})
```

#### Passing a custom ts-morph project (for the internal compiler)

(see [generate.ts](src/generate.ts) for the defaults)

```ts
import { Project } from 'ts-morph'
import { Generator } from '@runtyping/valibot'

const generator = new Generator({
  targetFile: 'src/runtypes.ts',
  project: new Project({
    // ...
  }),
})
```
