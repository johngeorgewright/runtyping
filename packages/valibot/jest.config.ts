import { Config } from 'jest'

const config: Config = {
  testEnvironment: 'node',
  transform: {
    '^.+\\.ts$': ['ts-jest', { tsconfig: '<rootDir>/test/tsconfig.json' }],
  },
}

export default config
