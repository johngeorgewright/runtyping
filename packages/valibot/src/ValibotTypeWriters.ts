import {
  DeclareType,
  Enum,
  getTypeName,
  Import,
  ImportFromSource,
  Static,
  sortUndefinedFirst,
  Tuple,
  TypeWriter,
  TypeWriters,
  Write,
} from '@runtyping/generator'
import { Type, ts } from 'ts-morph'

export default class ValibotTypeWriters extends TypeWriters {
  override *defaultStaticImplementation(type: Type): TypeWriter {
    yield [Import, { source: 'valibot', name: 'InferOutput' }]
    yield [Static, [type, 'InferOutput<typeof ${name}>']]
  }

  *#simple(
    name: string,
    {
      alias,
      type,
      typeWriter,
    }: { alias?: string; type?: Type; typeWriter?: TypeWriter } = {}
  ): TypeWriter {
    const importOptions = alias ? { alias } : {}
    yield [Import, { source: 'valibot', name, ...importOptions }]
    yield [Write, `${alias ?? name}(`]
    if (type) yield* this.generateOrReuseType(type)
    if (typeWriter) yield* typeWriter
    yield [Write, ')']
  }

  override null(): TypeWriter {
    return this.#simple('null', { alias: 'Null' })
  }

  override string(): TypeWriter {
    return this.#simple('string')
  }

  override number(): TypeWriter {
    return this.#simple('number')
  }

  override boolean(): TypeWriter {
    return this.#simple('boolean')
  }

  override undefined() {
    return this.#simple('undefined', { alias: 'Undefined' })
  }

  override unknown() {
    return this.#simple('unknown')
  }

  override void() {
    return this.#simple('void')
  }

  override never() {
    return this.#simple('never')
  }

  override any() {
    return this.#simple('any')
  }

  override array(_type: Type, elementType: Type): TypeWriter {
    return this.#simple('array', { type: elementType })
  }

  *#tuple(types: Type[]): TypeWriter {
    yield [Import, { source: 'valibot', name: 'strictTuple' }]
    yield [Write, 'strictTuple([']
    for (const element of types) {
      yield* this.generateOrReuseType(element)
      yield [Write, ', ']
    }
    yield [Write, '])']
  }

  override tuple(type: Type): TypeWriter {
    return this.#tuple(type.getTypeArguments())
  }

  override *variadicTuple(type: Type): TypeWriter {
    yield [Import, { source: 'valibot', name: 'array' }]
    yield [Import, { source: 'valibot', name: 'check' }]
    yield [Import, { source: 'valibot', name: 'minLength' }]
    yield [Import, { source: 'valibot', name: 'pipe' }]
    yield [Write, 'pipe(']
    yield* this.#simple('array', { typeWriter: this.unknown() })
    yield [Write, ', ']
    yield [Write, `minLength(${Tuple.getTupleMinSize(type)}), `]
    yield [Write, `check(data => `]
    yield* this.variadicTupleElements({
      tupleType: type,
      *element(type, index) {
        yield [Import, { source: 'valibot', name: 'is' }]
        yield [Write, 'is(']
        yield* this.generateOrReuseType(type)
        yield [
          Write,
          `, ${index >= 0 ? `data[${index}]` : `data[data.length${index}]`}`,
        ]
        yield [Write, ')']
      },
      *variadicElement(this: ValibotTypeWriters, type, from, to) {
        yield [Import, { source: 'valibot', name: 'is' }]
        yield [Import, { source: 'valibot', name: 'length' }]
        yield [Write, 'is(']
        yield [Write, 'pipe(']
        yield* this.#simple('array', { type })
        yield [Write, `, length(data.slice(${from}, ${to}).length)`]
        yield [Write, ')']
        yield [Write, `, data.slice(${from}, ${to})`]
        yield [Write, ')']
      },
      *separator() {
        yield [Write, ' && ']
      },
    })
    yield [Write, ')']
    yield [Write, ')']
  }

  override *union(type: Type): TypeWriter {
    const [first, ...rest] = type.getUnionTypes().sort(sortUndefinedFirst)

    if (!first) return yield* this.undefined()
    if (!rest.length) return yield* this.generateOrReuseType(first)

    yield [Import, { source: 'valibot', name: 'union' }]

    yield [Write, 'union([']
    yield* this.generateOrReuseType(first)
    for (const item of rest) {
      yield [Write, ', ']
      yield* this.generateOrReuseType(item)
    }
    yield [Write, '])']
  }

  override *withGenerics(typeWriter: TypeWriter, type: Type): TypeWriter {
    yield [Import, { source: 'valibot', name: 'InferOutput' }]
    yield [Import, { source: 'valibot', name: 'GenericSchema' }]
    const closeFunction = yield* this.openGenericFunction(
      type,
      'GenericSchema',
      'InferOutput'
    )
    yield* typeWriter
    yield* closeFunction()
  }

  override *enum(type: Type): TypeWriter {
    const name = getTypeName(type)
    const alias = `_${name}`
    yield [Import, { source: 'valibot', name: 'enum', alias: 'Enum' }]
    yield [ImportFromSource, { name, alias }]
    yield [Write, `Enum(${alias})`]
  }

  override *enumLiteral(type: Type): TypeWriter {
    const enumTypeName = Enum.getEnumIdentifierNameFromEnumLiteral(type)
    yield [ImportFromSource, { name: enumTypeName, alias: `_${enumTypeName}` }]
    yield* this.#literal(`_${enumTypeName}.${getTypeName(type)}`)
  }

  override *builtInObject(type: Type): TypeWriter {
    yield [Import, { source: 'valibot', name: 'instance' }]
    yield [Write, `instance(${type.getText()})`]
  }

  override *function(_type: Type): TypeWriter {
    yield [Import, { source: 'valibot', name: 'function', alias: 'func' }]
    yield [Write, 'func()']
  }

  override *intersection(type: Type): TypeWriter {
    const [first, ...rest] = type
      .getIntersectionTypes()
      .sort(sortUndefinedFirst)

    if (!first) return yield* this.undefined()
    if (!rest.length) return yield* this.generateOrReuseType(type)

    yield [Import, { source: 'valibot', name: 'intersect' }]

    yield [Write, 'intersect([']
    yield* this.generateOrReuseType(first)
    for (const item of rest) {
      yield [Write, ', ']
      yield* this.generateOrReuseType(item)
    }
    yield [Write, '])']
  }

  override *lazy(type: Type): TypeWriter {
    const name = getTypeName(type)
    const alias = `_${name}`
    yield [Import, { source: 'valibot', name: 'lazy' }]
    yield [Import, { source: 'valibot', name: 'GenericSchema' }]
    yield [ImportFromSource, { alias, name }]
    yield [DeclareType, `GenericSchema<${alias}>`]
    yield [Write, 'lazy(() => ']
    yield* this.typeWriter(type)
    yield [Write, ')']
  }

  override literal(type: Type): TypeWriter {
    return this.#literal(type.getText())
  }

  *#literal(literal: string): TypeWriter {
    yield [Import, { source: 'valibot', name: 'literal' }]
    yield [Write, `literal(${literal})`]
  }

  override *numberIndexedObject(type: Type): TypeWriter {
    yield [Import, { source: 'valibot', name: 'record' }]
    yield [Import, { source: 'valibot', name: 'number' }]
    yield [Write, 'record(number(), ']
    yield* this.generateOrReuseType(type.getNumberIndexType()!)
    yield [Write, ')']
  }

  override *object(type: Type<ts.ObjectType>): TypeWriter {
    yield [Import, { source: 'valibot', name: 'object' }]
    yield [Write, 'object({']
    yield* this.objectProperties(type, {
      *whenOptional(propertyWriter) {
        yield [Import, { source: 'valibot', name: 'optional' }]
        yield [Write, 'optional(']
        yield* propertyWriter
        yield [Write, ')']
      },
    })
    yield [Write, '})']
  }

  override *stringIndexedObject(type: Type): TypeWriter {
    yield [Import, { source: 'valibot', name: 'custom' }]
    yield [Import, { source: 'valibot', name: 'intersect' }]
    yield [Import, { source: 'valibot', name: 'record' }]
    yield [Import, { source: 'valibot', name: 'string' }]
    yield [Write, 'intersect([']
    yield [Write, 'record(string(), ']
    yield* this.generateOrReuseType(type.getStringIndexType()!)
    yield [Write, ')'] // /record
    yield [
      Write,
      ', custom((data) => !Array.isArray(data), "Unexpected array")',
    ]
    yield [Write, '])'] // /intersect
  }

  override *attachTransformer(
    typeWriter: TypeWriter,
    fileName: string,
    exportName: string
  ): TypeWriter {
    const alias = `${exportName}Transformer`
    yield [Import, { source: 'valibot', name: 'pipe' }]
    yield [Import, { source: 'valibot', name: 'transform' }]
    yield [Import, { source: fileName, name: exportName, alias }]
    yield [Write, 'pipe(']
    yield* typeWriter
    yield [Write, `, transform(${alias})`]
    yield [Write, `)`]
  }
}
