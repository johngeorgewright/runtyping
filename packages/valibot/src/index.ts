export { default as Generator } from './Generator'
export * from '@runtyping/generator/dist/runtypes'
