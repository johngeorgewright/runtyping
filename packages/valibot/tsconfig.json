{
  "compilerOptions": {
    "target": "es6",
    "module": "commonjs",
    "lib": ["es2018"],
    "importHelpers": true,
    "declaration": true,
    "sourceMap": true,
    "outDir": "./dist",
    "downlevelIteration": true,
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noImplicitOverride": true,
    "moduleResolution": "node",
    "esModuleInterop": true,
    "resolveJsonModule": true
  },
  "include": ["src"]
}
