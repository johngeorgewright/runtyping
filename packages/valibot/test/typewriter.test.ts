import testTypeWriter from '@runtyping/test-type-writers'
import { number, object, parse, string } from 'valibot'
import ValibotTypeWriters from '../src/ValibotTypeWriters'

testTypeWriter({
  createNumberValidator: number,
  createStringValidator: string,
  createObjectValidator: object,
  typeWriters: new ValibotTypeWriters(),
  validate: parse,
})
