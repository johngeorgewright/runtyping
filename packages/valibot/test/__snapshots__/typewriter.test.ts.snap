// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`without static types 1`] = `
"import { literal, number, object, string, union } from 'valibot';

// This file is generated by runtyping (https://github.com/johngeorgewright/runtyping).
// Manual changes might be lost - proceed with caution!
export const One = object({ type: literal("one"), value: string(), });
export const Two = object({ type: literal("two"), value: number(), });
export const Three = union([One, Two]);
"
`;
